### PR TITLE
1202 Write container build for Intel 21 OneAPI

### LINKED ISSUE #1202: Test compilation with Intel 21
**Describe the bug**
I have an image set up. Here's the error I get near the end:

```
[352/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_construct_no_idx.extended.cc.o
FAILED: tests/CMakeFiles/collection_extended.dir/unit/collection/test_construct_no_idx.extended.cc.o
/usr/bin/ccache /opt/intel/oneapi/mpi/2021.1.1/bin/mpicxx -DFMT_HEADER_ONLY=1 -DFMT_USE_USER_DEFINED_LITERALS=0 -DHAS_DETECTION_COMPONENT=1 -I/vt/tests/unit -I/vt/lib/fmt -I/vt/lib/CLI -I/vt/lib/libfort/lib -Irelease -I/vt/src -Ilib/checkpoint/src -I/vt/lib/checkpoint/src -I/vt/lib/detector/src -isystem /vt/tests/extern/googletest/googletest/include -isystem /vt/tests/extern/googletest/googletest -Wall -pedantic -Wshadow -Wno-unknown-pragmas -Wsign-compare -O3 -DNDEBUG -fPIC -qopenmp -std=c++14 -MD -MT tests/CMakeFiles/collection_extended.dir/unit/collection/test_construct_no_idx.extended.cc.o -MF tests/CMakeFiles/collection_extended.dir/unit/collection/test_construct_no_idx.extended.cc.o.d -o tests/CMakeFiles/collection_extended.dir/unit/collection/test_construct_no_idx.extended.cc.o -c /vt/tests/unit/collection/test_construct_no_idx.extended.cc
icpc: error #10106: Fatal error in /opt/intel/oneapi/compiler/2021.1.1/linux/bin/intel64/../../bin/intel64/mcpcom, terminated by kill signal
compilation aborted for /vt/tests/unit/collection/test_construct_no_idx.extended.cc (code 1)
[353/511] Building CXX object tests/CMakeFiles/runtime_basic.dir/unit/runtime/test_memory_footprint.cc.o
[354/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_insert.extended.cc.o
[355/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_lb_lite.extended.cc.o
/vt/tests/unit/collection/test_lb_lite.extended.cc(109): warning #3868: declaration hides member "vt::tests::unit::lb_lite::LBTest::val2" (declared at line 95) in same scope
    double val2 = 0.4f;
           ^

[356/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_lb.extended.cc.o
[357/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_model_per_collection.extended.cc.o
[358/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_collection_group.extended.cc.o
[359/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_checkpoint.extended.cc.o
[360/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_index_types.extended.cc.o
[361/511] Building CXX object tests/CMakeFiles/collection_extended.dir/unit/collection/test_broadcast.extended.cc.o
ninja: build stopped: subcommand failed.

real    18m21.599s
user    117m48.416s
sys     8m43.522s
The command '/bin/sh -c /vt/ci/build_cpp.sh /vt /build' returned a non-zero code: 1
```

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #1202

### CODE REVIEW

File: ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ARG proxy=""
+
+ENV https_proxy=${proxy} \
+    http_proxy=${proxy}
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+RUN apt-get update -y -q && \
+    apt-get install -y -q --no-install-recommends \
+    intel-oneapi-compiler-dpcpp-cpp \
```
Comment (cz4rs) [inline ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile:13]: TODO: check if this can be slimmed down further (using only `classic` package)

File: ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    VT_ASAN_ENABLED=${VT_ASAN_ENABLED} \
+    VT_POOL_ENABLED=${VT_POOL_ENABLED} \
+    VT_MPI_GUARD_ENABLED=${VT_MPI_GUARD_ENABLED} \
+    VT_EXTENDED_TESTS_ENABLED=${VT_EXTENDED_TESTS_ENABLED} \
+    VT_UNITY_BUILD_ENABLED=${VT_UNITY_BUILD_ENABLED} \
+    VT_DIAGNOSTICS_ENABLED=${VT_DIAGNOSTICS_ENABLED} \
+    VT_DIAGNOSTICS_RUNTIME_ENABLED=${VT_DIAGNOSTICS_RUNTIME_ENABLED} \
+    CMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
+
+ENV CCL_CONFIGURATION='cpu_gpu_dpcpp'
```
Comment (cz4rs) [inline ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile:73]: FIXME: this is copy-pasta from Intel's dockerfiles, remove unnecessary variables

Comment (cz4rs) [reply ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile]: I have removed the ones that were obviously obsolete.

File: ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile (hunk 1 above)
Comment (cz4rs) [reply ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile]: In the end, it has boiled down to installing following packages (+ dependencies):
```
    intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic \
    intel-oneapi-mpi-devel
```

**Comment by cz4rs** [comment]
To give some idea about the size and dependencies installed on top of `os-tools-ubuntu18.04`:
```
root@be71392eb44e:/# apt install --no-install-recommends intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic intel-oneapi-mpi-devel
Reading package lists... Done
(...)
The following NEW packages will be installed:
  intel-oneapi-common-licensing-2021.3.0 intel-oneapi-common-vars intel-oneapi-compiler-cpp-eclipse-cfg
  intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic
  intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic-2021.3.0
  intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic-common-2021.3.0
  intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic-runtime-2021.3.0
  intel-oneapi-compiler-dpcpp-cpp-classic-fortran-shared-runtime-2021.3.0
  intel-oneapi-compiler-dpcpp-cpp-common-2021.3.0 intel-oneapi-compiler-dpcpp-cpp-runtime-2021.3.0
  intel-oneapi-compiler-dpcpp-eclipse-cfg intel-oneapi-compiler-shared-2021.3.0
  intel-oneapi-compiler-shared-common-2021.3.0 intel-oneapi-compiler-shared-common-runtime-2021.3.0
  intel-oneapi-compiler-shared-runtime-2021.3.0 intel-oneapi-condaindex
  intel-oneapi-dev-utilities-2021.3.0 intel-oneapi-dev-utilities-eclipse-cfg
  intel-oneapi-dpcpp-cpp-2021.3.0 intel-oneapi-dpcpp-debugger-10.1.2
  intel-oneapi-dpcpp-debugger-eclipse-cfg intel-oneapi-mpi-2021.3.0 intel-oneapi-mpi-devel
  intel-oneapi-mpi-devel-2021.3.0 intel-oneapi-openmp-2021.3.0 intel-oneapi-openmp-common-2021.3.0
  intel-oneapi-tbb-2021.3.0 intel-oneapi-tbb-common-2021.3.0 intel-oneapi-tbb-common-devel-2021.3.0
  intel-oneapi-tbb-devel-2021.3.0
0 upgraded, 30 newly installed, 0 to remove and 6 not upgraded.
Need to get 1008 MB of archives.
After this operation, 3746 MB of additional disk space will be used.
```

The base image [(os-tools-ubuntu18.04)](https://github.com/intel/oneapi-containers/blob/master/images/docker/os-tools-ubuntu18.04/Dockerfile) is pretty lightweight, it just adds Intel repo and a couple of basic tools on top of ubuntu 18.04.

File: ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+ENV MPICH_CC=icc \
+    MPICH_CXX=icpc \
+    CC=mpicc \
+    CXX=mpicxx
+
+ENV MPI_EXTRA_FLAGS="" \
+    PATH=/usr/lib/ccache/:$PATH
+
+ENV CCL_CONFIGURATION='cpu_gpu_dpcpp' \
```
Comment (cz4rs) [inline ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile:40]: this configuration could probably be replaced with [setvars.sh setup](https://software.intel.com/content/www/us/en/develop/documentation/get-started-with-intel-oneapi-base-linux/top/before-you-begin.html#before-you-begin_GUID-338EB548-7DB6-410E-B4BF-E65C017389C4)

Comment (cz4rs) [reply ci/docker/ubuntu-18.04-intel-oneapi-cpp.dockerfile]: In the end, it seems like there is no good way to persist the environment between `RUN` instructions (other than using `ENV`). I have trimmed down the configuration and this is probably as good as it gets.

**Comment by cz4rs** [comment]
I have used `icc-18` pipeline to try out the new Dockerfile on Azure. This is using `icx/icpx` to build `mpich` and then `vt`.

There's only one test failing:
```
119/455 Test #119: vt:TestModelNorm.test_model_norm_3_no_mpi ..........................................***Failed  Error regular expression found in output. Regex=[FAILED]  0.01 sec
2021-07-26T12:40:07.4727770Z Note: Google Test filter = TestModelNorm.test_model_norm_3
2021-07-26T12:40:07.4728205Z [==========] Running 1 test from 1 test suite.
2021-07-26T12:40:07.4729187Z [----------] Global test environment set-up.
2021-07-26T12:40:07.4729786Z [----------] 1 test from TestModelNorm
2021-07-26T12:40:07.4730441Z [ RUN      ] TestModelNorm.test_model_norm_3
2021-07-26T12:40:07.4730911Z /vt/tests/unit/collection/test_model_norm.nompi.cc:208: Failure
2021-07-26T12:40:07.4731355Z Expected equality of these values:
2021-07-26T12:40:07.4731695Z   work_val
2021-07-26T12:40:07.4731996Z     Which is: 1
2021-07-26T12:40:07.4732768Z   expected_norms[obj.id - 1]
2021-07-26T12:40:07.4733125Z     Which is: 60
2021-07-26T12:40:07.4733529Z /vt/tests/unit/collection/test_model_norm.nompi.cc:208: Failure
2021-07-26T12:40:07.4733966Z Expected equality of these values:
2021-07-26T12:40:07.4734433Z   work_val
2021-07-26T12:40:07.4734720Z     Which is: 1
2021-07-26T12:40:07.4735382Z   expected_norms[obj.id - 1]
2021-07-26T12:40:07.4735738Z     Which is: 30
2021-07-26T12:40:07.4736120Z [  FAILED  ] TestModelNorm.test_model_norm_3 (1 ms)
2021-07-26T12:40:07.4736743Z [----------] 1 test from TestModelNorm (1 ms total)
```

**Comment by cz4rs** [comment]
**TLDR**
`std::isfinite` returns `true` for `std::numeric_limits<double>::infinity()` when using `icpx`

Reduced example @ Compiler Explorer: https://godbolt.org/z/KEqYs1e51

**Context**
We use `std::isfinite` in `Norm::getWork`. Failing test is using `std::numeric_limits<double>::infinity()` as the power value for computing the norm.

After adding following log:
```diff
--- a/src/vt/vrt/collection/balance/model/norm.cc
+++ b/src/vt/vrt/collection/balance/model/norm.cc
@@ -44,6 +44,7 @@
 
 #include "vt/vrt/collection/balance/model/norm.h"
 #include <cmath>
+#include <iostream>
 
 namespace vt { namespace vrt { namespace collection { namespace balance {
 
@@ -60,6 +61,7 @@ TimeType Norm::getWork(ElementIDStruct object, PhaseOffset offset)
   if (offset.subphase != PhaseOffset::WHOLE_PHASE)
     return ComposedModel::getWork(object, offset);
 
+  std::cout << "std::isfinite:\t" << std::boolalpha << std::isfinite(power_) << std::endl;
   if (std::isfinite(power_)) {
     double sum = 0.0;
```

Output using `icpx` (`Intel(R) oneAPI DPC++/C++ Compiler 2021.3.0 (2021.3.0.20210619)`):
```
root@7609992bd94c:/build# ./tests/collection_nompi --gtest_filter=TestModelNorm.test_model_norm_3
(...)
[ RUN      ] TestModelNorm.test_model_norm_3
std::isfinite:	true
/vt/tests/unit/collection/test_model_norm.nompi.cc:209: Failure
Expected equality of these values:
  work_val
    Which is: 1
  expected_norms[obj.id - 1]
    Which is: 60
std::isfinite:	true
/vt/tests/unit/collection/test_model_norm.nompi.cc:209: Failure
Expected equality of these values:
  work_val
    Which is: 1
  expected_norms[obj.id - 1]
    Which is: 30
[  FAILED  ] TestModelNorm.test_model_norm_3 (1 ms)
```

Using gcc / clang:
```
> ./tests/collection_nompi --gtest_filter=TestModelNorm.test_model_norm_3
(...)
[ RUN      ] TestModelNorm.test_model_norm_3
std::isfinite:  false
std::isfinite:  false
[       OK ] TestModelNorm.test_model_norm_3 (0 ms)
(...)
[  PASSED  ] 1 test.
```

**Comment by PhilMiller** [comment]
I'll report the compiler issue to the lab's point of contact, and I guess investigate a little further

**Comment by cz4rs** [comment]
Success after adding `-fhonor-infinites` flag:

image

[build.log](https://github.com/DARMA-tasking/vt/files/6947316/build.log)

**Comment by cz4rs** [comment]
Explanation:

> The problem is that icx's default floating point model is -fp-model fast
One of the things that -fp-model fast does it that it sets an attribute in the LLVM IR on floating point operations that says they can never yield infinity values.
So, when optimization is turned on, the LLVM IR ends up assuming that the call to `std::isfinite(power)` MUST return true, since the fp model says that inf values will never be generated.
You can use either `-fp-model precise` option, or `-fhonor-infinities` option to tell the compiler that it needs to assume infinities may be generated. Note that a similar problem can occur for NANs. fp-model fast assumes that NANs cannot occur/be generated, and `-fp-model precise` or `-fhonor-nans` will enable NANs .
```
$ icpx -O1 -fp-model precise test.cpp
$ ./a.out
false
$ icpx -O1 -fhonor-infinities test.cpp
$ ./a.out
false
```
> And it was in compiler version 2021.2 where the fp-model default was changed from precise to fast.
This was done purposefully for floating point model compatibility with the Intel classic compiler (ICC/ICL).
 
> the change in 2021.2 ICX was to make fp-model fast the default.  The behavior for NaNs and infinites is in the community LLVM code which we use as the base for our compiler.  The -fhonor and behaviors are from the community and not our additions to LLVM.  We are working with the community on a number of FP related changes we’d like to see from the LLVM community.

**Comment by cz4rs** [comment]
@lifflander 
I have added `icpx` workflow, compilation and tests pass fine locally. Can you please add a pipeline in Azure?

File: cmake-modules/SetCXXCompilerFlags.cmake [hunk 4]
```diff
@@ -24,6 +24,8 @@ elseif ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
   if (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 6)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-missing-braces")
   endif()
+elseif (CMAKE_CXX_COMPILER_ID STREQUAL IntelLLVM AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 2021.3.0)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fhonor-infinites")
```
Comment (PhilMiller) [inline cmake-modules/SetCXXCompilerFlags.cmake:28]: Given that this flag affects behavior of generated code, we should be very cautious about applying it, especially as a workaround for a bug. Though, at least as far as I can tell, it's not going to be inherited by dependent applications.

**Comment by lifflander** [comment]
@cz4rs I have the pipeline added to Azure now!

**Comment by cz4rs** [comment]
> @cz4rs I have the pipeline added to Azure now!

Great! I think it's good for merging as long as we get an approval from someone.

File: cmake-modules/SetCXXCompilerFlags.cmake (hunk 4 above)
Comment (cz4rs) [reply cmake-modules/SetCXXCompilerFlags.cmake:28]: A better option could be to use `target_compile_options` (which can specify `INTERFACE|PUBLIC|PRIVATE` scope), but we don't really use that in `vt` and it would require more changes.

**Review by JacobDomagala — APPROVED** [review]

File: cmake-modules/SetCXXCompilerFlags.cmake (hunk 4 above)
Comment (cz4rs) [reply cmake-modules/SetCXXCompilerFlags.cmake:28]: Overall, I think this should be fine as is, but we could create a separate issue for some CMake improvements.

**Comment by cz4rs** [comment]
TODO: use `target_compile_options` instead of global `CMAKE_CXX_FLAGS`

**Comment by jstrzebonski** [comment]
@cz4rs There are still compilation errors, example:
```bash
FAILED: src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx.o 
/usr/bin/ccache /usr/local/bin/mpicxx -DJSON_USE_IMPLICIT_CONVERSIONS=1 -I/vt/lib/CLI -Irelease -I/vt/src -I/vt/lib/json/include -I/vt/lib/brotli/c/include -I/vt/lib/libfort/lib -isystem /vt/lib/fmt/include -isystem /build/checkpoint/install/include -isystem /build/detector/install/include -Wall -pedantic -Wshadow -Wno-unknown-pragmas -Wsign-compare -ftemplate-backtrace-limit=100 -fhonor-infinites -Werror -O3 -DNDEBUG -fPIC -std=c++14 -MD -MT src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx.o -MF src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx.o.d -o src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx.o -c src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx
In file included from src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx:17:
/vt/src/vt/vrt/collection/balance/model/norm.cc:54:16: error: comparison with NaN always evaluates to false in fast floating point modes [-Werror,-Wtautological-constant-compare]
  vtAssert(not std::isnan(power), "Power must have a real value");
               ^~~~~~~~~~~~~~~~~
/vt/src/vt/configs/error/config_assert.h:156:40: note: expanded from macro 'vtAssert'
    vtAssertImpl(vt_assert_should_fail,cond,str)
                                       ^~~~
/vt/src/vt/configs/error/config_assert.h:126:13: note: expanded from macro 'vtAssertImpl'
      if (!(cond)) {                                                    \
            ^~~~
1 error generated.
```

**Review by jstrzebonski — CHANGES_REQUESTED** [review]
@cz4rs There are still compilation errors, example:
```bash
FAILED: src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx.o 
/usr/bin/ccache /usr/local/bin/mpicxx -DJSON_USE_IMPLICIT_CONVERSIONS=1 -I/vt/lib/CLI -Irelease -I/vt/src -I/vt/lib/json/include -I/vt/lib/brotli/c/include -I/vt/lib/libfort/lib -isystem /vt/lib/fmt/include -isystem /build/checkpoint/install/include -isystem /build/detector/install/include -Wall -pedantic -Wshadow -Wno-unknown-pragmas -Wsign-compare -ftemplate-backtrace-limit=100 -fhonor-infinites -Werror -O3 -DNDEBUG -fPIC -std=c++14 -MD -MT src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx.o -MF src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx.o.d -o src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx.o -c src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx
In file included from src/CMakeFiles/vt.dir/Unity/unity_12_cxx.cxx:17:
/vt/src/vt/vrt/collection/balance/model/norm.cc:54:16: error: comparison with NaN always evaluates to false in fast floating point modes [-Werror,-Wtautological-constant-compare]
  vtAssert(not std::isnan(power), "Power must have a real value");
               ^~~~~~~~~~~~~~~~~
/vt/src/vt/configs/error/config_assert.h:156:40: note: expanded from macro 'vtAssert'
    vtAssertImpl(vt_assert_should_fail,cond,str)
                                       ^~~~
/vt/src/vt/configs/error/config_assert.h:126:13: note: expanded from macro 'vtAssertImpl'
      if (!(cond)) {                                                    \
            ^~~~
1 error generated.
```

**Comment by jstrzebonski** [comment]
BTW, you could also add pipeline badge to readme.

File: cmake-modules/SetCXXCompilerFlags.cmake (hunk 4 above)
Comment (cz4rs) [reply cmake-modules/SetCXXCompilerFlags.cmake:28]: New issue created to handle this in all `vt`: https://github.com/DARMA-tasking/vt/issues/1602.

**Comment by lifflander** [comment]
Approved... but I can't actually approve it since I'm an author.
